### ISSUE TITLE
Redshift continuously fails

### ISSUE THREAD

**Issue by Kendos-Kenlen** [NixOS/nixpkgs#23363]
## Issue description

When Redshift is enabled with the latitude and longitude specified, it spam logs with the following error messages:

```Mar 02 01:09:07 gauthier-nixos systemd[1925]: redshift.service: Service hold-off time over, scheduling restart.
Mar 02 01:09:07 gauthier-nixos systemd[1925]: Stopped Redshift colour temperature adjuster.
Mar 02 01:09:07 gauthier-nixos systemd[1925]: Started Redshift colour temperature adjuster.
Mar 02 01:09:07 gauthier-nixos redshift[9786]: No protocol specified
Mar 02 01:09:07 gauthier-nixos redshift[9786]: `RANDR Query Version' returned error -1
Mar 02 01:09:07 gauthier-nixos redshift[9786]: Initialization of randr failed.
Mar 02 01:09:07 gauthier-nixos redshift[9786]: Trying next method...
Mar 02 01:09:07 gauthier-nixos redshift[9786]: No protocol specified
Mar 02 01:09:07 gauthier-nixos redshift[9786]: X request failed: XOpenDisplay
Mar 02 01:09:07 gauthier-nixos redshift[9786]: Initialization of vidmode failed.
Mar 02 01:09:07 gauthier-nixos redshift[9786]: Trying next method...
Mar 02 01:09:07 gauthier-nixos redshift[9786]: No more methods to try.
Mar 02 01:09:07 gauthier-nixos systemd[1925]: redshift.service: Main process exited, code=exited, status=1/FAILURE
Mar 02 01:09:07 gauthier-nixos systemd[1925]: redshift.service: Unit entered failed state.
Mar 02 01:09:07 gauthier-nixos systemd[1925]: redshift.service: Failed with result 'exit-code'.
```

However, Redshift work as my screen is yellowed and nothing looks wrong about it.

### Steps to reproduce

Enable Redshift and specify latitude & longitude. This is my config:

```  services.redshift.enable = true;
  services.redshift.latitude = "47.21806";
  services.redshift.longitude = "-1.55278";
```

## Technical details

* System: 17.03pre101839.53a2baa (Gorilla)
* Nix version: nix-env (Nix) 1.11.6
* Nixpkgs version: 17.03pre101839.53a2baa

**Comment by peterhoeg** [comment]
It's because redshift is ```wantedBy = [ "default.target" ];``` which means it's being spawned for the ```sddm``` user as well causing this problem.

The proper way to deal with this is to have sddm launch the various graphical environments using systemd targets that then pull in redshift.

You can work around it like this: ```wantedBy = lib.mkForce [ "graphical-session.target" ];``` but you then need another target to manually activate that pulls this in.

Bottom-line: it will stay broken for a while until proper targets are in place for graphical environments.

It's on my list to fix.

**Comment by michaelpj** [comment]
See https://github.com/NixOS/nixpkgs/issues/22670

**Comment by Kendos-Kenlen** [comment]
Thank you for the answer.
I think the best way now is to install and use redshift in the user profile instead of the system profile, avoiding the error for now.

**Comment by peterhoeg** [comment]
>   I think the best way now is to install and use redshift in the user
>   profile instead of the system profile, avoiding the error for now.

It is being run by the systemd user instance - it's just being run by
the wrong target.

**Comment by rnhmjoj** [comment]
redshift is already a user service and there is no `graphical.target` for those:
```
$ systemctl --user list-units --type target
UNIT           LOAD   ACTIVE SUB    DESCRIPTION                                                                                          
basic.target   loaded active active Basic System                                                                                         
default.target loaded active active Default                                                                                              
paths.target   loaded active active Paths                                                                                                
sockets.target loaded active active Sockets                                                                                              
timers.target  loaded active active Timers 
```
Currently all the these services are implemented like this: `default.target` and `Restart=always`. I don't think there is any other way.

**Comment by peterhoeg** [comment]
With systemd 232 we have ```graphical-session.target``` and ```graphical-session-pre.target``` but those are not being used yet. That's the proper place to put redshift.

**Comment by peterhoeg** [comment]
What I'm doing here in the meantime is to define a ```kde-session.target``` like this:

```
[Unit]
BindsTo=graphical-session.target
Description=KDE
Requires=kwalletd5.service xembedsniproxy.service
```

As I also use redshift, that is then ```wantedBy = lib.mkForce [ "graphical-session.target" ];``` as mentioned above.

I then trigger the starting of kde-session.target using an alias after logging in.

It's super ugly, but it works as an interim solution.

It's

**Comment by dtzWill** [comment]
FWIW this impacts the compton service as well (which exhibits the same symptoms of constantly failing to start for other users).

**Comment by cmacrae** [comment]
Simply commenting to confirm that this does indeed also impact the compton service in the same manner :+1:

**Comment by fpletz** [comment]
This won't be fixed in 17.03 but is fixed in master.

**Comment by hugbubby** [comment]
> This won't be fixed in 17.03 but is fixed in master.

This is exact problem is happening right now for me:

```
May 10 19:47:13 lcnix systemd[1265]: Started Redshift colour temperature adjuster.
May 10 19:47:13 lcnix redshift[8433]: `RANDR Query Version' returned error -1
May 10 19:47:13 lcnix redshift[8433]: Initialization of randr failed.
May 10 19:47:13 lcnix redshift[8433]: Trying next method...
May 10 19:47:13 lcnix redshift[8433]: X request failed: XOpenDisplay
May 10 19:47:13 lcnix redshift[8433]: Initialization of vidmode failed.
May 10 19:47:13 lcnix redshift[8433]: Trying next method...
May 10 19:47:13 lcnix redshift[8433]: No more methods to try.
May 10 19:47:13 lcnix systemd[1265]: redshift.service: Main process exited, code=exited, status=1/FAILURE
May 10 19:47:13 lcnix systemd[1265]: redshift.service: Failed with result 'exit-code'.

```
